content/posts/2025-01-30-ebpf-flows-filtering.md

```diff
@@ -0,0 +1,229 @@
+| tcpFlags    | TCP flags to filter flows by                             | "SYN";"SYN-ACK";"ACK";"FIN";"RST";"URG";"ECE";"CWR";"FIN-ACK";"RST-ACK" |           |
+
+
+Note:
+
+- A rule cannot include both ports and either sourcePorts or destPorts simultaneously.
+- The number of eBPF flow filter rules is capped at `16` to maintain efficient memory usage when this feature is enabled.
+- Validation webhook rejects rules containing duplicate CIDRs.
+- Supports both IPv4 and IPv6 formats.
+- If users wish to match any CIDR, the default `0.0.0.0/0` (Null CIDR) can be used.
+
+The example below demonstrates various filtering options using multi-rule filters.
+
+```yaml
+agent:
+  type: eBPF
+  ebpf:
+    flowFilter:
+      enable: true
+      rules:
+        - action: Accept
+          cidr: 10.128.0.0/24
+          peerCIDR: 10.129.0.0/24
+          ports: '443,6443'
+          protocol: TCP
+          sampling: 10
+        - action: Accept
+          cidr: 10.129.0.1/24
+          ports: 53
+          protocol: UDP
+          sampling: 20
+        - action: Reject
+          tcpFlags: SYN
+          cidr: 10.130.0.0/24
+          protocol: TCP
+          sourcePorts: 80-100
+        - action: Accept
+          cidr: 172.30.0.0/16
+          protocol: SCTP
+          pktDrops: true
+        - action: Reject
+          cidr: 8.8.8.8/32
+          protocol: ICMP
+          icmpType: 8 // ICMP Echo request packet
+```
+2️⃣ Packet Inspection – Extract relevant packet attributes within an eBPF program.
+
+3️⃣ Early Flow Filtering –
+Skip or Allow packets based on predefined conditions before doing further ebpf packets processing.
+The Following Diagram shows how eBPF filtering is done
+
+<img src="{page.image('ebpf-flows-filtering/ebpf-flows-filtering.png')}" alt="eBPF Flows Filtering Processing">
+
+## Key Use Cases
+
+🚀 Reducing Observability Overhead – Avoid logging irrelevant flows in high-traffic Kubernetes clusters.
+
+🔐 Security Filtering – Focus on anomalous or suspicious traffic while ignoring normal flows.
+
+🌐 Network Performance Monitoring – Capture only high-latency or dropped-packet flows for troubleshooting.
+
+### Filter EastWest and NorthSouth flows
+
+Allow traffic to service IP 172.30.100.64:80 with a sampling rate of 10.
+Permit communication between pods in the 10.128.0.0/16 and 10.129.0.0/16 subnets with a sampling rate of 20.
+Reject all other traffic.
+
+```yaml
+agent:
+  type: eBPF
+  ebpf:
+    flowFilter:
+      enable: true
+      rules:
+        - action: Accept
+          cidr: 172.30.0.0/16
+          sampling: 10
+        - action: Accept
+          cidr: 10.128.0.0/16
+          peerCIDR: 10.129.0.0/16
+          sampling: 20
+        - action: Reject
+          cidr: 0.0.0.0/0
+``` 
+
+<img src="{page.image('ebpf-flows-filtering/ebpf-svc-and-pods-flows.png')}" alt="eBPF Flows Filtering Kubernetes NorthSouth and EastWest Flows">
+
```

Comment:
You could also mention the other use-case where users would on the contrary want to have 1:1 sampling on North-South (which was what we've been asked to allow, so we know it's something users might be interested in). Something like:
If on the contrary, you're more interested in North-South traffic but still without loosing visibility on East-West, you can simply edit the last rule as:

```
        - action: Accept
          cidr: 0.0.0.0/0
          sampling: 1
```

Reply:
won't that be a special case of the general case I used in my example ?

Reply:
If I read your example correctly, they're 2 specific and different use cases:
- your example shows keeping East-West and rejecting all North-South
- my suggestion is to show keeping both North-South and East-West, but with higher sampling defined on East-West

... unless if you consider traffic to 172.30.0.0/16 being the North-South traffic? But I don't think it's right, since Service IPs are also used in E-W (services configured as ClusterIP)

I'd rather define as "North-South" anything that has any IP neither in 10.128.0.0/14 nor in 172.30.0.0/16

For instance, a pod 10.128.0.1 that establishes a connection to an outside service like 8.8.8.8 should be categorized in N-S, right?

Reply:
I added additional case for 8.8.8.8 with different config options lmk if that is what u were looking for

Reply:
Actually I was thinking of it as something part of the same E-W / N-S use case (it's just that I had a broader definition of N-S than what you're showing)

If I had to distinguish E-W / N-S traffic with different sampling, I'd do it that way:

```
      rules:
        - action: Accept
          cidr: 0.0.0.0/0
          sampling: 1
        - action: Accept
          cidr: 10.128.0.0/14
          peerCIDR: 10.128.0.0/14
          sampling: 50
        - action: Accept
          cidr: 172.30.0.0/16
          peerCIDR: 10.128.0.0/14
          sampling: 50
```

So that would use sampling 50 for pod-to-pod and pod-to-service, and 1 for everything else, including external-to-service or pod-to-external.

Reply:
this rule open things up more than 8.8.8.8 I would like to open things up only for 8.8.8.8 ping to explore the APIs options I will merge my two usecases

Reply:
sure, thanks